coapthon/messages/message.py

```diff
@@ -454,7 +454,8 @@ def etag(self, etag):
         for e in etag:
             option = Option()
             option.number = defines.OptionRegistry.ETAG.number
-            if not  isinstance(e, bytes):
+            if isinstance(e, str):
```

Comment:
Can there be an assert that we expect `bytes` otherwise?

Reply:
We allow it to be `None` or `bytes` otherwise. Same as in _CoAPthon2_ and it is necessary for our test cases. I'm not exactly sure about `e` being `int` though, so I would like to keep it as is for now.

Reply:
The "allowance" is not made explicit. This is what I meant.

By `if not  isinstance(e, bytes):`, it was expecting bytes and trying converting everything else.
By changing it to ` if isinstance(e, str):`, this acceptance criteria is not as clear as before I would say. Therefore, IMO, it could be interesting to catch any other types and abort.

In the end, I'm also fine to keep it as is.

Reply:
Statements like `if isinstance(e, str):` try to fix specific difference between Python2 and 3 about string handling. In Python2 `str` and bytes sequences were essentially the same thing and, therefore didn't require any action neither from original library author, nor from tado users to distinguish when they meant encoded data on wire and when some ASCII-string payload. With Python3 strings are `str` and encoded data is `bytes` which has to be explicitly converted between each other.

I think that `if not is instance (e, bytes):` is about this `bytes`->`str` conversion fix as it looks exactly the same in many other places where CoAPthon2 and 3 diverge. Giacomo Tanganelli just didn't take into account that tag could also be `None`.

Reply:
Would it harm to have a statement like this in place?
```py
assert isinstance(e, (bytes, bytearray))
```

Reply:
I'll add to my _develop_ branch for investigation